aptos-move/aptos-vm-logging/src/lib.rs

```diff
@@ -0,0 +1,179 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+pub mod counters;
+pub mod log_schema;
+
+pub mod prelude {
+    pub use crate::{
+        counters::CRITICAL_ERRORS, speculative_debug, speculative_error, speculative_info,
+        speculative_log, speculative_trace, speculative_warn,
+    };
+}
+
+use crate::{counters::CRITICAL_ERRORS, log_schema::AdapterLogSchema};
+use aptos_logger::{prelude::*, Level};
+use aptos_speculative_state_helper::{SpeculativeEvent, SpeculativeEvents};
+use arc_swap::ArcSwapOption;
+use once_cell::sync::Lazy;
+use std::sync::Arc;
+
+struct VMLogEntry {
+    level: Level,
+    context: AdapterLogSchema,
+    message: String,
+}
+
+impl VMLogEntry {
+    fn new(level: Level, context: AdapterLogSchema, message: String) -> Self {
+        Self {
+            level,
+            context,
+            message,
+        }
+    }
+}
+
+// Implement dispatching for a recorded VM log entry to support recording
+// as speculative logging event (and eventual dispatching as needed).
+impl SpeculativeEvent for VMLogEntry {
+    fn dispatch(self) {
+        match self.level {
+            Level::Error => {
+                // TODO: Consider using SpeculativeCounter to increase CRITICAL_ERRORS
+                // on the critical path instead of async dispatching.
+                error!(self.context, "{}", self.message);
+                CRITICAL_ERRORS.inc();
+            },
+            Level::Warn => warn!(self.context, "{}", self.message),
+            Level::Info => info!(self.context, "{}", self.message),
+            Level::Debug => debug!(self.context, "{}", self.message),
+            Level::Trace => trace!(self.context, "{}", self.message),
+        }
+    }
+}
+
+static BUFFERED_LOG_EVENTS: Lazy<ArcSwapOption<SpeculativeEvents<VMLogEntry>>> =
+    Lazy::new(|| ArcSwapOption::from(None));
+
+/// Initializes the storage of speculative logs for num_txns many transactions.
+pub fn init_speculative_logs(num_txns: usize) {
+    BUFFERED_LOG_EVENTS.swap(Some(Arc::new(SpeculativeEvents::new(num_txns))));
+}
+
+/// Adds a message at a specified logging level and given context (that includes txn index)
+/// to speculative buffer. Logs directly and logs a separate (new error) if the speculative
+/// events storage is not initialized or appropriately sized.
+pub fn speculative_log(level: Level, context: &AdapterLogSchema, message: String) {
+    let txn_idx = context.txn_idx;
+    match &*BUFFERED_LOG_EVENTS.load() {
+        Some(log_events) => {
+            let log_event = VMLogEntry::new(level, context.clone(), message);
+            if let Err(e) = log_events.record(txn_idx, log_event) {
+                error!("{:?}", e);
+                CRITICAL_ERRORS.inc();
```

Comment:
this pattern (error! + counter) seems repeated a lot, probably can have something combined them

Reply:
done, created alert! macro